MINOR: doc change for minisr to clarify replicas in Kafka Config

pdruley opened the pull request:
This is a minor doc change to MinInSyncReplicasDoc to clarify what is a replica and further help users understand the importance of this configuration in relation to their producer configuration.

It didn't seem like a Jira was necessary but I am open to creating one if needed. I did not compile from source to test, no other code was changed.

This contribution is my original work and I license the work to the project under the project's open source license.

Thanks!

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

pdruley committed:
MINOR: doc change for minisr

joel-hamill commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -826,10 +826,11 @@ object KafkaConfig {
     "create a topic with a replication factor of 3, set min.insync.replicas to 2, and " +
     "produce with acks of \"all\". This will ensure that the producer raises an exception " +
-    "if a majority of replicas do not receive a write."
+    "if a majority of replicas (2) do not receive a write."
```
```suggestion
    "if a majority of replicas, in this case 2, do not receive a write."
```

joel-hamill commented:
LGTM, had one comment

pdruley committed:
Update core/src/main/scala/kafka/server/KafkaConfig.scala

Co-authored-by: Joel Hamill <11722533+joel-hamill@users.noreply.github.com>

divijvaidya commented:
@pdruley can you please fix the conflict and we can work towards merging this PR.

pdruley committed:
Merge branch 'trunk' into pdruley/minisrDocChange

pdruley commented:
> @pdruley can you please fix the conflict and we can work towards merging this PR.

I resolved the conflict but where are the docs for the configuration settings now?

TaiJuWu committed:
MINOR: restore `testGetAllTopicMetadataShouldNotCreateTopicOrReturnUnknownTopicPartition` (#18633)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-18592 Cleanup ReplicaManager (#18621)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Christo Lolov <lolovc@amazon.com>, Chia-Ping Tsai <chia7712@gmail.com>

divijvaidya commented:
> > @pdruley can you please fix the conflict and we can work towards merging this PR.
> 
> I resolved the conflict but where are the docs for the configuration settings now?

Here: https://github.com/apache/kafka/blob/40890faa1bed8296919a7e108937ec559d901348/server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java#L138

mjsax committed:
MINOR: cleanup ProcessorContextImplTest (#18682)

Reviewers: Bill Bejeck <bill@confluent.io>

frankvicky committed:
KAFKA-18620: Remove UnifiedLog#legacyFetchOffsetsBefore (#18686)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

LoganZhuZzz committed:
KAFKA-18597 Fix max-buffer-utilization-percent is always 0 (#18627)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-18559 Cleanup FinalizedFeatures (#18593)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-18590 Cleanup DelegationTokenManager (#18618)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

ocadaruma committed:
KAFKA-16372 Fix producer doc discrepancy with the exception behavior (#15574)

Currently, Producer.send doc is inconsistent with actual exception behavior
      - TimeoutException: This won't be thrown from send on buffer-full or metadata-missing actually. Instead, it will returned as failed future.
       - AuthenticationException/AuthorizationException: These exceptions are also won't be thrown. Returned with failed future actually.
Fixed Callback javadoc and ProducerConfig doc as well.

Reviewers: Luke Chen <showuon@gmail.com>, Andrew Schofield <aschofield@confluent.io>

dajac committed:
KAFKA-18616; Refactor DumpLogSegments's MessageParsers (#18688)

All the work that we have done to automate and to simplify the coordinator records allows us to simplify the related MessageParsers in DumpLogSegments. They can all share the same based implementation. This is nice because it ensures that we handle all those records similarly.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

m1a2st committed:
KAFKA-18474: Remove zkBroker listener (#18477)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>, PoAn Yang <payang@apache.org>

mumrah committed:
KAFKA-17587 Refactor test infrastructure (#18602)

This patch reorganizes our test infrastructure into three Gradle modules:

":test-common:test-common-internal-api" is now a minimal dependency which exposes interfaces and annotations only. It has one project dependency on server-common to expose commonly used data classes (MetadataVersion, Feature, etc). Since this pulls in server-common, this module is Java 17+. It cannot be used by ":clients" or other Java 11 modules.

":test-common:test-common-util" includes the auto-quarantined JUnit extension. The @Flaky annotation has been moved here. Since this module has no project dependencies, we can add it to the Java 11 list so that ":clients" and others can utilize the @Flaky annotation

":test-common:test-common-runtime" now includes all of the test infrastructure code (TestKitNodes, etc). This module carries heavy dependencies (core, etc) and so it should not normally be included as a compile-time dependency.

In addition to this reorganization, this patch leverages JUnit SPI service discovery so that modules can utilize the integration test framework without depending on ":core". This will allow us to start moving integration tests out of core and into the appropriate sub-module. This is done by adding ":test-common:test-common-runtime" as a testRuntimeOnly dependency rather than as a testImplementation dependency. A trivial example was added to QuorumControllerTest to illustrate this.

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-18630: Clean ReplicaManagerBuilder (#18687)

Reviewers: Christo Lolov <lolovc@amazon.com>

apoorvmittal10 committed:
KAFKA-17894: Implemented broker topic metrics for Share Group 1/N (KIP-1103) (#18444)

The PR implements the BrokerTopicMetrics defined in KIP-1103.

The PR also corrected the share-acknowledgement-rate and share-acknowledgement-count metrics defined in KIP-932 as they are moved to BrokerTopicMetrics, necessary changes to KIP-932 broker metrics will be done once we complete KIP-1103.

Reviewers: Andrew Schofield <aschofield@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>, Jun Rao <junrao@gmail.com>

mjsax committed:
KAFKA-13722: remove usage of old ProcessorContext  (#18292)

We want to deprecate an remove the old ProcessorContext. Thus, we need
to refactor Kafka Streams runtime code, to not make calls into the old
ProcessorContext but only into new code path.

Reviewers: Bill Bejeck <bill@confluent.io>

CalvinLiu7947 committed:
KAFKA-16540: Clear ELRs when min.insync.replicas is changed. (#18148)

In order to maintain the integrity of replication, we need to clear the ELRs of affected partitions when min.insync.replicas is changed. This could happen at the topic level, or at a global level if the cluster level default is changed.

Reviewers: Colin P. McCabe <cmccabe@apache.org>

TaiJuWu committed:
KAFKA-18529: ConsumerRebootstrapTest should run for async consumer (#18554)

Reviewers: Kirk True <ktrue@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>, Lianet Magrans <lmagrans@confluent.io>

mingdaoy committed:
KAFKA-18499 Clean up zookeeper from LogConfig (#18583)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mingyen066 committed:
KAFKA-18639 Enable the @Flaky annotation for some flaky tests (#18701)

The following tests were previously reported as flaky but were only annotated with a comment in pull request #18558 due to module dependency limitations:

    testAdminClientApisAuthenticationFailure
    testOutdatedCoordinatorAssignment
    testThrottledProducerConsumer

With the introduction of the new test infrastructure #18602 , which allows all modules to use the @Flaky annotation, these tests should now be updated to include the @Flaky annotation.

Reviewers: TengYao Chi <kitingiao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

m1a2st committed:
KAFKA-18533 Remove KafkaConfig zookeeper related logic (#18547)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
MINOR: add removed controller metrics to zk2kraft (#18654)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
KAFKA-18555 Avoid casting MetadataCache to KRaftMetadataCache (#18632)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

chirag-wadhwa5 committed:
KAFKA-18642: Increased the timeouts in share_consumer_test.py system tests (#18699)

Reviewers: Andrew Schofield <aschofield@confluent.io>

frankvicky committed:
MINOR: Remove EMPTY_NODE_ID (#18707)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mingyen066 committed:
KAFKA-18631 Remove ZkConfigs (#18693)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

smjn committed:
KAFKA-18632: Added few share consumer multibroker tests. (#18679)

Reviewers: Andrew Schofield <aschofield@confluent.io>

msillence committed:
KAFKA-17792: Efficiently parse decimals with large exponents in Connect Values (#17510)

Reviewers: Greg Harris <greg.harris@aiven.io>, Mickael Maison <mickael.maison@gmail.com>

mjsax committed:
KAFKA-18541: fix flaky KafkaStreamsTelemetryIntegrationTest (#18569)

Reviewers: Bill Bejeck <bill@confluent.io>

TaiJuWu committed:
KAFKA-18528: MultipleListenersWithSameSecurityProtocolBaseTest and GssapiAuthenticationTest should run for async consumer (#18555)

Reviewers: Kirk True <ktrue@confluent.io>, Lianet Magrans <lmagrans@confluent.io>

mjsax committed:
KAFKA-17162: join() started thread in DefaultTaskManagerTest (#18570)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

apoorvmittal10 committed:
KAFKA-18657: Fixing SharePartitionManager flaky test (#18710)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Andrew Schofield <aschofield@confluent.io>

m1a2st committed:
MINOR: Remove ZooKeeper mentions in comments (#18646)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

smjn committed:
KAFKA-18629: Delete share group state impl [1/N] (#18712)

Reviewers: Christo Lolov <lolovc@amazon.com>, Andrew Schofield <aschofield@confluent.io>

apoorvmittal10 committed:
KAFKA-17951: Share parition rotate strategy (#18651)

Reviewers: Andrew Schofield <aschofield@confluent.io>, Abhinav Dixit <adixit@confluent.io>

mumrah committed:
MINOR Mark a StickyAssignorTest as flaky (#18719)

Mark StickyAssignorTest#testLargeAssignmentAndGroupWithNonEqualSubscription as flaky. Used data from this
report https://github.com/apache/kafka/actions/runs/12982945953

Reviewers: Andrew Schofield <aschofield@confluent.io>

ijuma committed:
KAFKA-18648: Add back support for metadata version 0-3 (#18716)

During testing, we identified that kafka-python (and aiokafka) relies on metadata request v0 and
hence we need to add these back to comply with the premise of KIP-896 - i.e. it should not
break the clients listed within it.

I reverted the changes from #18218 related to the removal of metadata versions 0-3.

I will submit a separate PR to undeprecate these API versions on the relevant 3.x branches.

kafka-python (and aiokafka) work correctly (produce & consume) with this change on
top of the 4.0 branch.

Reviewers: David Arthur <mumrah@gmail.com>

AndrewJSchofield committed:
KAFKA-18488: Improve KafkaShareConsumerTest (#18728)

Reviewers: Lianet Magrans <lmagrans@confluent.io>

frankvicky committed:
KAFKA-18619: New consumer topic metadata events should set requireMetadata flag (#18668)

Reviewers: Lianet Magrans <lmagrans@confluent.io>

ijuma committed:
KAFKA-18646: Null records in fetch response breaks librdkafka (#18726)

Ensure we always return empty records (including cases where an error is returned).
We also remove `nullable` from `records` since it is effectively expected to be
non-null by a large percentage of clients in the wild.

This behavior regressed in fe56fc9 (KAFKA-18269). Empty records were
previously set via `FetchResponse.recordsOrFail(partitionData)` in the
now-removed `maybeConvertFetchedData` method.

Added an integration test that fails without this fix and also update many
tests to set `records` to `empty` instead of leaving them as `null`.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, David Arthur <mumrah@gmail.com>

adixitconfluent committed:
KAFKA-18653: Fix mocks and potential thread leak issues causing silent RejectedExecutionException in share group broker tests (#18725)

Reviewers: Apoorv Mittal <apoorvmittal10@gmail.com>, Andrew Schofield <aschofield@confluent.io>

jeffkbkim committed:
MINOR: prevent exception from HdrHistogram (#18674)

HdrHistogram can throw an exception if the recorded value is greater than a configured limit. Expand the ceiling from per-metric to all invocations.

Reviewers: David Jacot <djacot@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

smjn committed:
KAFKA-18632: Multibroker test improvements. (#18718)

Reviewers: Andrew Schofield <aschofield@confluent.io>

CalvinLiu7947 committed:
KAFKA-18662: Return CONCURRENT_TRANSACTIONS on produce request in TV2 (#18733)

While testing, it was found that the not_enough_replicas error was super common and could be easily confused. Since we are already bumping the request, we can signify that the produce request may return this error and new clients can handle it 

(Note, the java client should be able to handle this already as a retriable error, but other client libraries may need to implement this change)

Reviewers: Justine Olshan <jolshan@confluent.io>

FrankYang0529 committed:
KAFKA-18383 Remove reserved.broker.max.id and broker.id.generation.enable (#18478)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

fonsdant committed:
KAFKA-16339: Add Kafka Streams migrating guide from transform to process (#18314)

Reviewers: Matthias J. Sax <matthias@confluent.io>

bbejeck committed:
KAFKA-18498: Update lock ownership from main thread (#18732)

Once a StreamThread receives its assignment, it will close the startup tasks. But during the closing process, the StandbyTask.closeClean() method will eventually call theStatemanagerUtil.closeStateManager method which needs to lock the state directory, but locking requires the calling thread be the current owner. Since the main thread grabs the lock on startup but moves on without releasing it, we need to update ownership explicitly here in order for the stream thread to close the startup task and begin processing.

Reviewers: Matthias Sax <mjsax@apache.org>, Nick Telford

mjsax committed:
KAFKA-13722: remove internal usage of old ProcessorContext (#18698)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

frankvicky committed:
KAFKA-18569: New consumer close may wait on unneeded FindCoordinator (#18590)

Reviewers: Lianet Magrans <lmagrans@confluent.io>, Kirk True <ktrue@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

CalvinLiu7947 committed:
KAFKA-18649: complete ClearElrRecord handling (#18708)

Implement ClearElrRecord handling in the TopicDelta. Also, the ReplicationControlManager should not merge updates if ELR/LastKnownElr are empty, becuase that will cause an unnecessary partition epoch bump.

Reviewers: Colin P. McCabe <cmccabe@apache.org>

mumrah committed:
KAFKA-18636 Fix how we handle Gradle exits in CI (#18681)

This patch removes the explicit failure of test tasks in Gradle when there is a flaky test. This also fixes a fall-through case in junit.py where we did not recognize an error prior to running the tests (such as the javadoc task).

Additionally, this patch removes usages of ignoreFailures in our CI and changes the XML copy task to a finalizer task instead of doLast closure.

Reviewers: Jun Rao <junrao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
KAFKA-18441: Fix flaky KafkaAdminClientTest#testAdminClientApisAuthenticationFailure (#18735)

Reviewers: Lianet Magrans <lmagrans@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>, Andrew Schofield <aschofield@confluent.io>

lucasbru committed:
KAFKA-18621: Add StreamsCoordinatorRecordHelpers (#18669)

A class with helper methods to create records stored in the __consumer_offsets topic.

Compared to the feature branch, I added unit tests (most functions were not tested) and adopted the new interface for constructing coordinator records introduced by David.

Reviewers: Bruno Cadonna <cadonna@apache.org>

mehbey committed:
KAFKA-18570: Update documentation to add remainingLogsToRecover, remainingSegmentsToRecover and LogDirectoryOffline metrics (#18731)

Reviewers: Divij Vaidya <diviv@amazon.com>

agavra committed:
MINOR: fix typo in HTML docs (#18742)

Reviewers: Divij Vaidya <diviv@amazon.com>

yx9o committed:
MINOR: Improve error message for invalid topic in TopicCommand (#18714)

Reviewers: Divij Vaidya <diviv@amazon.com>

pramithas committed:
MINOR: KafkaProducerTest - Fix resource leakage and replace explicit invocation of close() method with try with resources (#18678)

Reviewers: Divij Vaidya <diviv@amazon.com>, Greg Harris <greg.harris@aiven.io>, Christo Lolov <lolovc@amazon.com>

smjn committed:
MINOR: Refactor share coord cache helper to share package. (#18743)

Reviewers: Andrew Schofield <aschofield@confluent.io>

m1a2st committed:
KAFKA-18034: CommitRequestManager should fail pending requests on fatal coordinator errors (#18548)

Reviewers: Lianet Magrans <lmagrans@confluent.io>, Kirk True <ktrue@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (1/N) - filter[Not]/selectKey (#18703)

Reviewers: Alieh Saeedi <asaeedi@confluent.io>, Lucas Brutschy <lbrutschy@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (2/N) - print/foreach/peek/split/merge (#18704)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

fonsdant committed:
MINOR: remove close from contextual processors javadoc

Reviewers: Matthias J. Sax <matthias@confluent.io>

kirktrue committed:
KAFKA-17182: Consumer fetch sessions are evicted too quickly with AsyncKafkaConsumer (#17700)

This change reduces fetch session cache evictions on the broker for AsyncKafkaConsumer by altering its logic to determine which partitions it includes in fetch requests.

Background
Consumer implementations fetch data from the cluster and temporarily buffer it in memory until the user next calls Consumer.poll(). When a fetch request is being generated, partitions that already have buffered data are not included in the fetch request.

The ClassicKafkaConsumer performs much of its fetch logic and network I/O in the application thread. On poll(), if there is any locally-buffered data, the ClassicKafkaConsumer does not fetch any new data and simply returns the buffered data to the user from poll().

On the other hand, the AsyncKafkaConsumer consumer splits its logic and network I/O between two threads, which results in a potential race condition during fetch. The AsyncKafkaConsumer also checks for buffered data on its application thread. If it finds there is none, it signals the background thread to create a fetch request. However, it's possible for the background thread to receive data from a previous fetch and buffer it before the fetch request logic starts. When that occurs, as the background thread creates a new fetch request, it skips any buffered data, which has the unintended result that those partitions get added to the fetch request's "to remove" set. This signals to the broker to remove those partitions from its internal cache.

This issue is technically possible in the ClassicKafkaConsumer too, since the heartbeat thread performs network I/O in addition to the application thread. However, because of the frequency at which the AsyncKafkaConsumer's background thread runs, it is ~100x more likely to happen.

Options
The core decision is: what should the background thread do if it is asked to create a fetch request and it discovers there's buffered data. There were multiple proposals to address this issue in the AsyncKafkaConsumer. Among them are:

The background thread should omit buffered partitions from the fetch request as before (this is the existing behavior)
The background thread should skip the fetch request generation entirely if there are any buffered partitions
The background thread should include buffered partitions in the fetch request, but use a small “max bytes” value
The background thread should skip fetching from the nodes that have buffered partitions
Option 4 won out. The change is localized to AbstractFetch where the basic idea is to skip fetch requests to a given node if that node is the leader for buffered data. By preventing a fetch request from being sent to that node, it won't have any "holes" where the buffered partitions should be.

Reviewers: Lianet Magrans <lmagrans@confluent.io>, Jeff Kim <jeff.kim@confluent.io>, Jun Rao <junrao@gmail.com>

jolshan committed:
KAFKA-18660: Transactions Version 2 doesn't handle epoch overflow correctly (#18730)

Fixed the typo that used the wrong producer ID and epoch when returning so that we handle epoch overflow correctly.

We also had to rearrange the concurrent transaction handling so that we don't self-fence when we start the new transaction with the new producer ID.

I also tested this with a modified version of the code where epoch overflow happens on the first epoch bump (every request has a new producer id)

Reviewers: Artem Livshits <alivshits@confluent.io>, Jeff Kim <jeff.kim@confluent.io>

kevin-wu24 committed:
KAFKA-16524; Metrics for KIP-853 (#18304)

This change implement some of the metrics enumerated in KIP-853.

The KafkaRaftMetrics object now exposes number-of-voters, number-of-observers and uncommitted-voter-change. The number-of-observers and uncommitted-voter-change metrics are only present on the active controller or leader, since it does not make sense for other replicas to report these metrics.

In order to make these two metrics thread-safe, KafkaRaftMetrics needs to be passed into LeaderState, and therefore QuorumState. This introduces a circularity since the KafkaRaftMetrics constructor takes in QuorumState. To break the circularity for now, the logic using QuorumState will be moved to the KafkaRaftMetrics#initialize method.

The BrokerServerMetrics object now exposes ignored-static-voters. The ControllerServerMetrics object now exposes IgnoredStaticVoters. To implement both metrics for "ignored static voters", this PR introduces the ExternalKRaftMetrics interface, which allows for higher layer metrics objects to be accessible within the raft module.

Reviewers: José Armando García Sancio <jsancio@apache.org>

mjsax committed:
KAFKA-18644: improve generic type names for KStreamImpl and KTableImpl (#18722)

Reviewers: Bill Bejeck <bill@confluent.io>, Lucas Brutschy <lbrutschy@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (3/N) - groupBy[Key] (#18705)

Reviewers: Alieh Saeedi <asaeedi@confluent.io>, Lucas Brutschy <lbrutschy@confluent.io>

showuon committed:
MINOR: Clean up for sasl endpoints (#18519)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

mimaison committed:
KAFKA-15995: Initial API + make Producer/Consumer plugins Monitorable (#17511)


Reviewers: Greg Harris <gharris1727@gmail.com>, Luke Chen <showuon@gmail.com>

frankvicky committed:
KAFKA-18677; Update ConsoleConsumerTest system test (#18763)

This patch converts the ConsoleConsumerTest system test to only use KRaft.

Reviewers: David Jacot <djacot@confluent.io>

dajac committed:
KAFKA-18146; tests/kafkatest/tests/core/upgrade_test.py needs to be re-added as KRaft (#18766)

This patch renames kraft_upgrade_test.py to upgrade_test.py. This is enough to cover the old upgrade/downgrade tests.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

dajac committed:
KAFKA-18320; Ensure that assignors are at the right place (#18750)

The full class name of the assignors if part of our public api. Hence, we should ensure that they are not changed by mistake. This patch adds a unit test verifying them.

Reviewers: Sean Quah <squah@confluent.io>, Jeff Kim <jeff.kim@confluent.io>

lianetm committed:
Revert "KAFKA-17182: Consumer fetch sessions are evicted too quickly with AsyncKafkaConsumer (#17700)"

This reverts commit 6cf54c4dab9ff39f21d7f2c9bd241b938b5651b2.

apoorvmittal10 committed:
KAFKA-18683: Handle slicing of file records for updated start position (#18759)

The PR corrects the check which was introduced in #5332 where position is checked to be within boundaries of file. The check 
    position > currentSizeInBytes - start 
is incorrect, since the position is relative to start.

Reviewers: Jun Rao <junrao@gmail.com>

ijuma committed:
KAFKA-18659: librdkafka compressed produce fails unless api versions returns produce v0 (#18727)

Return produce v0-v2 as supported versions in `ApiVersionsResponse`, but disable support
for it everywhere else.

Since clients pick the highest supported version by both client and broker during version
negotiation, this solves the problem with minimal tech debt (even though it's not ideal that
`ApiVersionsResponse` becomes inconsistent with the actual protocol support).

Add one test for the socket server handling (in `ProcessorTest`) and one test for the
client behavior (in `ProduceRequestTest`). Adjust a couple of api versions tests to verify
the new behavior.

Finally, include a few clean-ups in `ApiKeys`, `Protocol`, `ProduceRequest`,
`ProduceRequestTest` and `BrokerApiVersionsCommandTest`.

Reference to related librdkafka issue:
https://github.com/confluentinc/librdkafka/issues/4956

Reviewers: Jun Rao <junrao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>, Stanislav Kozlovski <stanislav_kozlovski@outlook.com>

LoganZhuZzz committed:
MINOR: Replace lambda expressions with method references in ProducerStateManager (#18753)

Reviewers: TengYao Chi <kitingiao@gmail.com>, Divij Vaidya <diviv@amazon.com>, Chia-Ping Tsai <chia7712@gmail.com>

mingyen066 committed:
KAFKA-18674 Document the incompatible changes in parsing --bootstrap-server (#18751)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
KAFKA-18678 Update TestVerifiableProducer system test (#18768)

Reviewers: David Jacot <djacot@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

aliehsaeedii committed:
KAFKA-17125: Streams Sticky Task Assignor (#18652)

Implements streams sticky assignor on the broker-side.

Reviewers: Bill Bejeck <bbejeck@apache.org>, Lucas Brutschy <lbrutschy@confluent.io>

dajac committed:
KAFKA-18672; CoordinatorRecordSerde must validate value version (#18749)

CoordinatorRecordSerde does not validate the version of the value to check whether the version is supported by the current version of the software. This is problematic if a future and unsupported version of the record is read by an older version of the software because it would misinterpret the bytes. Hence CoordinatorRecordSerde must throw an error if the version is unknown. This is also consistent with the handling in the old coordinator.

Reviewers: Jeff Kim <jeff.kim@confluent.io>

Yunyung committed:
MINOR: Remove the test for ZooKeeper metrics used by ZooKeeperClient (#18775)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
KAFKA-17631 Convert SaslApiVersionsRequestTest to kraft (#18330)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
KAFKA-18676; Update Benchmark system tests (#18785)

Update `benchmark_test.py` to use KRaft.

```
> TC_PATHS="tests/kafkatest/benchmarks/core/benchmark_test.py" /bin/bash tests/docker/run_tests.sh

================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.12.0
session_id:       2025-02-03--001
run time:         96 minutes 48.900 seconds
tests run:        120
passed:           120
flaky:            0
failed:           0
ignored:          0
================================================================================
```

Reviewers: David Jacot <djacot@confluent.io>

m1a2st committed:
KAFKA-18685: Cleanup DynamicLogConfig constructor (#18764)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Christo Lolov <lolovc@amazon.com>

dongnuo123 committed:
KAFKA-18655: Implement the consumer group size counter with scheduled task (#18717)

During testing we discovered that the empty group count is not updated in group conversion, but when the new group is transition to other state, the empty group count is decremented. This could result in negative empty group count.

We can have a new consumer group count implementation that follows the pattern we did for the classic group count. The timeout task periodically refreshes the metrics based on the current groups soft state.

Reviewers: Jeff Kim <jeff.kim@confluent.io>

lucasbru committed:
KAFKA-18325: Add TargetAssignmentBuilder (#18676)

A class to build a new target assignment based on the provided parameters. As a result, it yields the records that must be persisted to the log and the new member assignments as a map.

Compared to the feature branch, I extended the unit tests (testing also standby and warm-up task logic) and adopted simplifications due to the TasksTuple class.

Reviewers: Bruno Cadonna <cadonna@apache.org>, Bill Bejeck <bbejeck@apache.org>

m1a2st committed:
KAFKA-18545: Remove Zookeeper logic from LogManager (#18592)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Mickael Maison <mickael.maison@gmail.com>

kamalcph committed:
MINOR: Remove the noisy log in consumer manager (#18787)

The statement gets logged in the INFO level and gets printed for every message produced to the __remote_log_metadata topic. Removed the log statement as it is needed only during debug session. And, we have another log at DEBUG level to capture this information.

Reviewers: Luke Chen <showuon@gmail.com>, Christo Lolov <lolovc@amazon.com>

jolshan committed:
KAFKA-18654 [1/2]: Transaction Version 2 performance regression due to early return (#18720)

https://issues.apache.org/jira/browse/KAFKA-18575 solved a critical race condition by returning with CONCURRENT_TRANSACTIONS early when the transaction was still completing.
In testing, it was discovered that this early return could cause performance regressions.

Prior to KIP-890 the addpartitions call was a separate call from the producer. There was a previous change https://issues.apache.org/jira/browse/KAFKA-5477 that decreased the retry backoff to 20ms. With KIP-890 and making the call through the produce path, we go back to the default retry backoff which takes longer. Prior to 18575 we introduce a slight delay when sending to the coordinator, so prior to 18575, we are less likely to return quickly and get stuck in this backoff. However, based on results from produce benchmarks, we can still run into the default backoff in some scenarios.

This PR reverts KAFKA-18575, and doesn't return early and wait until the coordinator for checking if a transaction is ongoing. Instead, it will fix the handling with the verification guard so we don't hit the edge condition.

Also cleans up some of the verification text that was unclear.

Reviewers: Jeff Kim <jeff.kim@confluent.io>, Artem Livshits <alivshits@confluent.io>

mjsax committed:
MINOR: fix NPE in KS `Topology` for new `AutoOffsetReset` (#18780)

Introduced via KIP-1106.

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

mjsax committed:
MINOR: Session windows should accept zero as session gap (#18734)

Reviewers: Almog Gavra <almog@responsive.dev>, Anna Sophie Blee-Goldman <ableegoldman@apache.org>

mjsax committed:
MINOR: cleanup KStream JavaDocs (4/N) - stream-table-inner-join (#18721)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>, Bill Bejeck <bill@confluent.io>

mingyen066 committed:
KAFKA-18693 Remove PasswordEncoder (#18790)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mjsax committed:
KAFKA-18644: improve generic type names for internal FK-join classes (#18700)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

CalvinLiu7947 committed:
KAFKA-18635: reenable the unclean shutdown detection (#18277)

We need to re-enable the unclean shutdown detection when in ELR mode, which was inadvertently removed during the development process.

Reviewers: David Mao <dmao@confluent.io>,  Jun Rao <junrao@gmail.com>

mingyen066 committed:
MINOR: Correct the link in the Javadoc for test-common-internal-api (#18788)

Reviewers: Ken Huang <s7133700@gmail.com>, TengYao Chi <kitingiao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

alozano3 committed:
MINOR: Add missing MirrorMaker2 metrics to docs (#18691)


Reviewers: Mickael Maison <mickael.maison@gmail.com>
Co-authored-by: u0184996 <alozano@caixabanktech.com>

dajac committed:
MINOR: Fix Benchmark sys tests (#18796)

This patch fixes the Benchmark system tests. We misconfigured the quorum in https://github.com/apache/kafka/commit/bc7b87001bfd641f93105eaa1e0326bdaf15b55d.

```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.12.0
session_id:       2025-02-04--001
run time:         57 minutes 27.169 seconds
tests run:        62
passed:           62
flaky:            0
failed:           0
ignored:          0
================================================================================
```

Reviewers: PoAn Yang <payang@apache.org>, Christo Lolov <lolovc@amazon.com>

dajac committed:
MINOR: Fix PerformanceService sys test (#18797)

This patch fixes the PerformanceService system test which was still using ZK.

```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.12.0
session_id:       2025-02-04--003
run time:         1 minute 42.629 seconds
tests run:        4
passed:           4                                                                                                                                                                         flaky:            0
failed:           0                                                                                                                                                                         ignored:          0
================================================================================
```

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

dajac committed:
MINOR: Fix TestBounce sys test (#18798)

```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.12.0
session_id:       2025-02-04--005
run time:         4 minutes 0.023 seconds
tests run:        4
passed:           4
flaky:            0
failed:           0
ignored:          0
================================================================================
```

Reviewers: Lianet Magrans <lmagrans@confluent.io>

mumrah committed:
MINOR increase max flaky tests allowed (#18792)

Increase the maximum number of flaky tests we tolerate for the main test suite from 3 to 10. This will result in fewer failed builds.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

showuon committed:
KAFKA-18230: Handle not controller or not leader error in admin client (#18165)


Reviewers: Mickael Maison <mickael.maison@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

cadonna committed:
KAFKA-18538: Add Streams membership manager (#18551)

The Streams membership manager is used client-side in the
background thread of the async consumer. For each member
/consumer, it is responsible for:
* keeping the member state,
* keeping assignments for the member,
* reconciling the assignments of the member -- for example
when tasks need to be revoked before other tasks are assigned
* requesting invocations of assignment and revocation callbacks
by the stream thread.

The Streams membership manager is called by the background thread of
the async consumer, directly in its event loop and from the
 Streams group heartbeat request manager. The Streams membership
manager uses the Streams rebalance events processor to request
assignment/revocation callback in the stream thread.

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>, Bill Bejeck <bill@confluent.io>

jolshan committed:
KAFKA-18691: Flaky test testFencingOnTransactionExpiration (#18793)

It appears this test was failing because the transaction was never aborting and the concurrent transactions errors would not go away.

ccab9eb introduced the test failure because it requires the transaction to complete, but I suspect the lack of completion was happening before the change.

The timeout for the write is based on the transactional timeout, and 100ms seemed too small -- thus the requests to update the state would often repeatedly time out.

Also removed the loop since it was not necessary.

Reviewers: Jeff Kim <jeff.kim@confluent.io>, Calvin Liu <caliu@confluent.io>

mjsax committed:
MINOR: disable "processing threads" in SmokeTestDriverIntegrationTest (#18773)

Reviewers: Bruno Cadonna <bruno@confluent.io>

squah-confluent committed:
KAFKA-18690: Keep leader metadata for RE2J-assigned partitions (#18777)

Reviewers: Lianet Magrans <lmagrans@confluent.io>

jolshan committed:
MINOR: Add missing test tag to UnifiedLogTest.scala (#18794)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mjsax committed:
MINOR: cleanup KStream JavaDocs (6/N) - map[Values] (#18755)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

mjsax committed:
HOTFIX: compilation error

Two merged PRs overlapped in a non-conflicting way, breaking compilation:
 - https://github.com/apache/kafka/pull/18722
 - https://github.com/apache/kafka/pull/18755

mingyen066 committed:
KAFKA-18675 Add tests for valid and invalid broker addresses (#18781)

Reviewers: Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-18698: Migrate suitable classes to records in server and server-common modules (#18783)

Reviewers: Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>, Christo Lolov <lolovc@amazon.com>

FrankYang0529 committed:
KAFKA-18705: Move ConfigRepository to metadata module (#18784)

Reviewers: TengYao Chi <kitingiao@gmail.com>, Christo Lolov <lolovc@amazon.com>

Rancho-7 committed:
KAFKA-18694: Migrate suitable classes to records in coordinator-common module (#18782)

Reviewers: TengYao Chi <kitingiao@gmail.com>, Ken Huang <s7133700@gmail.com>, Christo Lolov <lolovc@amazon.com>

kamalcph committed:
KAFKA-18722: Remove the unreferenced methods in TBRLMM and ConsumerManager (#18791)

Reviewers: Luke Chen <showuon@gmail.com>, Christo Lolov <lolovc@amazon.com>

frankvicky committed:
KAFKA-18645: New consumer should align close timeout handling with classic consumer (#18702)

Reviewers: Lianet Magrans <lmagrans@confluent.io>, Kirk True <ktrue@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

sjhajharia committed:
[KAFKA-16720] AdminClient Support for ListShareGroupOffsets (2/2) (#18671)

Reviewers: Apoorv Mittal <apoorvmittal10@gmail.com>, Sushant Mahajan <smahajan@confluent.io>, Andrew Schofield <aschofield@confluent.io>

smjn committed:
KAFKA-18629: Add persister impl and tests for DeleteShareGroupState RPC. [2/N] (#18748)

Reviewers: Andrew Schofield <aschofield@confluent.io>

chirag-wadhwa5 committed:
KAFKA-18494-3: solution for the bug relating to gaps in the share partition cachedStates post initialization (#18696)

Reviewers: Apoorv Mittal <apoorvmittal10@gmail.com>, Abhinav Dixit <adixit@confluent.io>, Andrew Schofield <aschofield@confluent.io>

lucasbru committed:
KAFKA-17379: Fix inexpected state transition from ERROR to PENDING_SHUTDOWN (#18765)

The exception stack trace shown in the the ticket can happen when we are
concurrently closing the producer because of an error and doing a
regular close. This is not a bug in the test, but a real race condition
that can happen.

The sequence is this:

Thread1: Enter PENDING_ERROR
Thread2: Check if state is already ERROR
Thread1: Transition to ERROR
Thread2: Check if state is already PENDING_ERROR
Thread2: Transition to PENDING_SHUTDOWN

One idea to fix this would be to synchronize the sequence performed by
Thread1 using the state lock. However, this would require more changes,
since we cannot use the normal state transition method `setState` while
owning the lock, as it calls user-defined callbacks, which may create
deadlocks. Do avoid adding more synchronization, we can also fix it by
first attempting to transition to PENDING_SHUTDOWN, and _then_ checking
whether another thread is already attempting to shut down (states
PENDING_SHUTDOWN, PENDING_ERROR, ERROR, NOT_RUNNING). Since we never
transition from a shutdown state back to a non-shutdown state.

Reviewers: Matthias J. Sax <matthias@confluent.io>

brandboat committed:
KAFKA-18206: EmbeddedKafkaCluster must set features (#18189)

related to KAFKA-18206, set features in EmbeddedKafkaCluster in both streams and connect module, note that this PR also fix potential transaction with empty records in sendPrivileged method as transaction version 2 doesn't allow this kind of scenario.

Reviewers: Justine Olshan <jolshan@confluent.io>

apoorvmittal10 committed:
MINOR: Removing share module from settings (#18806)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

fonsdant committed:
MINOR: Add javadocs to AbstractMergedSortedCacheStoreIterator (#18772)

While reviewing PR #18287, I wrote some javadocs to help me understand the AbstractMergedSortedCacheStoreIterator. Maybe we could add them to help the next developers getting into it.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

mjsax committed:
MINOR: update Kafka Streams `Topology` JavaDocs (#18778)

Reviewers: Bill Bejeck <bill@confluent.io>

stevenschlansker committed:
KAFKA-18689: Improve metric calculation to avoid NoSuchElementException (#18771)

Reviewers: Nick Telford <nick.telford@gmail.com>, Matthias J. Sax <matthias@confluent.io>

shahbazaamir committed:
MINOR: removed unwanted line breaks (#18744)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mingyen066 committed:
MINOR: Fix wrong config property in KafkaConfigTest (#18815)

Reviewers: Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

m1a2st committed:
KAFKA-18530 Remove ZooKeeperInternals (#18641)

Since zk has been removed in 4.0, config handlers no longer need to handle the "<default>" value. This PR streamlines the config update process by eliminating the unnecessary string checks for "<default>"

Reviewers: Christo Lolov <lolovc@amazon.com>, Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

cmccabe committed:
KAFKA-18360 Remove zookeeper configurations (#18566)

Remove broker.id.generation.enable and reserved.broker.max.id, which are not used in KRaft mode.
Remove inter.broker.protocol.version, which is not used in KRaft mode.

Reviewers: PoAn Yang <payang@apache.org>, Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

AndrewJSchofield committed:
KAFKA-18618: Improve leader change handling of acknowledgements [1/N] (#18672)

Reviewers: Apoorv Mittal <apoorvmittal10@gmail.com>, ShivsundarR <shr@confluent.io>, Manikumar Reddy <manikumar.reddy@gmail.com>

ppkarwasz committed:
KAFKA-18483 Disable `Log4jController` and `Loggers` if Log4j Core absent (#18496)

If Log4j Core is absent, most calls to Log4jController and Loggers will end up with a NoClassDefFoundError.

This changeset:

- Profits from the major version bump to rename k.util.Log4jController to LoggingController.
- Removes o.a.l.l.Level from the signature of public methods of o.a.k.connect.runtime.Loggers and replaces it with String.
- Provides an additional no-op implementation of k.util.LoggingController and o.a.k.connect.runtime.Loggers: if Log4j Core is not present on the runtime classpath the no-op implementation will be used.

Reviewers: Mickael Maison <mickael.maison@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

Rancho-7 committed:
KAFKA-18741 document the removal of `inter.broker.protocol.version` (#18818)

Reviewers: Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
MINOR: add missing <li> to upgrade.html (#18817)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mjsax committed:
MINOR: cleanup KStream JavaDocs (7/N) - repartition/to/toTable (#18760)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (8/N) - stream-stream-inner-join (#18761)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

peterxcli committed:
KAFKA-18275 Restarting broker in testing should use the same port (#18381)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

ijuma committed:
MINOR: Remove 'disable caching' workaround for Java 8 in connect (#18823)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Greg Harris <greg.harris@aiven.io>

FrankYang0529 committed:
KAFKA-17833: Convert DescribeAuthorizedOperationsTest to use KRaft (#18252)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

smjn committed:
KAFKA-18745: Handle network related errors in persister. (#18824)

Reviewers: Andrew Schofield <aschofield@confluent.io>

lucasbru committed:
MINOR: Fix streams smoke test flush records (#18830)

In the streams smoke test, flush records that are appended to the input topics, to advance the stream time so that all suppressed windows are flushed at the end of the test. The records are created with record time equal to current time + 2 days. caf0b67 changed the broker defaults so that records more than one hour in the future are rejected by the broker. This breaks the flush messages. By moving all record time stamps 2 days into the past, the existing logic should work correctly with the new default broker configuration.

A similar thing happens in the relational smoke test, where data is emitted 4 days into the future. To avoid running into retention / compaction, the window retention time is increased for both tests.

Reviewers: Bruno Cadonna <bruno@confluent.io>, Bill Bejeck <bill@confluent.io>

appchemist committed:
MINOR: Refactor method name for consistency (#18779)

Reviewers: Andrew Schofield <aschofield@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (5/N) - stream-globalTable-inner-join (#18747)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (9/N) - flatMap[Values] (#18805)

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (10/N) - stream-table-left-join (#18813)

Reviewers: Bill Bejeck <bill@confluent.io>

dajac committed:
MINOR: Accept specifying consumer group assignors by their short names (#18832)

At the moment, we require specifying builtin server side assignors by their full class name. This is not convenient and also exposed their full class name as part of our public API. This patch changes it to accept specifying builtin server side assignor by their short name (uniform or range) while continuing to accept full class name for customer assignors.

Reviewers: Jeff Kim <jeff.kim@confluent.io>

chirag-wadhwa5 committed:
KAFKA-18763: changed the assertion statement for acknowledgements to include only successful acks (#18846)

Reviewers: Andrew Schofield <aschofield@confluent.io>

brandboat committed:
MINOR: fix KStream#to incorrect javadoc (#18838)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
KAFKA-18441: Remove flaky tag on KafkaAdminClientTest#testAdminClientApisAuthenticationFailure (#18847)

Signed-off-by: PoAn Yang <payang@apache.org>
Reviewers: Andrew Schofield <aschofield@confluent.io>

m1a2st committed:
KAFKA-18225 ClientQuotaCallback#updateClusterMetadata is unsupported by kraft (#18196)

This commit ensures that the ClientQuotaCallback#updateClusterMetadata method is executed in KRaft mode. This method is triggered whenever a topic or cluster metadata change occurs. However, in KRaft mode, the current implementation of the updateClusterMetadata API is inefficient due to the requirement of creating a full Cluster object. To address this, a follow-up issue (KAFKA-18239) has been created to explore more efficient mechanisms for providing cluster information to the ClientQuotaCallback without incurring the overhead of a full Cluster object creation.

Reviewers: Mickael Maison <mickael.maison@gmail.com>, TaiJuWu <tjwu1217@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

Yunyung committed:
KAFKA-18743 Remove leader.imbalance.per.broker.percentage as it is not supported by Kraft (#18821)

Remove `leader.imbalance.per.broker.percentage` from config.
Add `leader.imbalance.per.broker.percentage` to release note

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

ijuma committed:
MINOR: Java version and TLS documentation improvements (#18822)

Most of the changes are obvious clean-ups/fixes. A couple of noteworthy items:

1. Support for non LTS versions is clarified (we were incorrectly stating full support
for Java 23).
2. TLS version negotiation details are clarified.

Reviewers: Matthias J. Sax <matthias@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

gongxuanzhang committed:
KAFKA-18658 add import control for examples module (#18812)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

m1a2st committed:
KAFKA-18366 Remove KafkaConfig.interBrokerProtocolVersion (#18820)

Reviewers: Ismael Juma <ismael@juma.me.uk>, Chia-Ping Tsai <chia7712@gmail.com>

TaiJuWu committed:
KAFKA-18770 close the RM created by testDelayedShareFetchPurgatoryOperationExpiration (#18853)

it's crucial to utilize a try-finally block to ensure proper closure of the ReplicaManager. Failing to do so can result in an unreleased thread from the purgatory, potentially leading to errors in subsequent integration tests that incorporate thread leak detection.

Reviewers: poorv Mittal <apoorvmittal10@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-18396: Migrate log4j1 configuration to log4j2 in KafkaDockerWrapper (#18394)

After log4j migration, we need to update the logging configuration in KafkaDockerWrapper from log4j1 to log4j2.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

mimaison committed:
KAFKA-14484: Move UnifiedLog static methods to storage (#18039)


Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

smjn committed:
KAFKA-18764: Throttle on share state RPCs auth failure. (#18855)

Reviewers: Andrew Schofield <aschofield@confluent.io>

edoardocomar committed:
KAFKA-18758:  NullPointerException in shutdown following InvalidConfigurationException (#18833)

* KAFKA-18758:  NullPointerException in shutdown following InvalidConfigurationException

Add checks for null in shutdown as BrokerLifecycleManager is not instantiaited if LogManager constructor throws an Exception

dajac committed:
MINOR: Fix log compaction system test (#18857)

`log.segment.bytes` must be greater or equals to 1MB (KIP-1030).

```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.12.0
session_id:       2025-02-10--003
run time:         55.903 seconds
tests run:        1
passed:           1
flaky:            0
failed:           0
ignored:          0
================================================================================
```

Reviewers: Divij Vaidya <diviv@amazon.com>

dajac committed:
MINOR: Fix reassign partitions system test (#18860)

The tests which set reassign_from_offset_zero=False have a setup phase which produces records with old timestamps to the topic and waits until they are cleaned by the retention in order to run the main phase of the test based on non-zero offsets. The setup phases did not wait enough for the cleaning task to kick in, mainly because the scheduled task was not started yet due to log.initial.task.delay.ms being set to 30s by default. Reducing it to 5s helps to stabilize the test. The patch also changes the sleep to 12s in order to have a bit more head room.

```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.12.0
session_id:       2025-02-11--016
run time:         26 minutes 9.451 seconds
tests run:        12
passed:           12
flaky:            0
failed:           0
ignored:          0
================================================================================
```

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

ShivsundarR committed:
Fix bug in json naming (#18858)

Reviewers: Apoorv Mittal <apoorvmittal10@gmail.com>, Andrew Schofield <aschofield@confluent.io>

chirag-wadhwa5 committed:
KAFKA-18756: Enabled share group configs for queues related system tests (#18829)

Reviewers: Andrew Schofield <aschofield@confluent.io>

mimaison committed:
KAFKA-15995: Adding KIP-877 support to Connect  (#17804)


Reviewers: Greg Harris <gharris1727@gmail.com>

mjsax committed:
MINOR: cleanup KStream JavaDocs (11/N) - stream-stream-left-join (#18836)

Reviewers: Bill Bejeck <bill@confluent.io>

mjsax committed:
MINOR: cleanup KStream JavaDocs (12/N) - process[Values] (#18839)

Reviewers: Bill Bejeck <bill@confluent.io>

jolshan committed:
KAFKA-18035: TransactionsTest testBumpTransactionalEpochWithTV2Disabled failed on trunk (#18451)

Sometimes we didn't get into abortable state before aborting, so the epoch didn't get bumped. Now we force abortable state with an attempt to send before aborting so the epoch bump occurs as expected.

Reviewers: Jeff Kim <jeff.kim@confluent.io>

FrankYang0529 committed:
KAFKA-18771: fix Flaky test KRaftClusterTest .testDescribeQuorumRequestToControllers (#18859)

The case testDescribeQuorumRequestToControllers shutdowns raft client but not the controller. This makes client has chance to send a request to the controller and get NOT_LEADER_OR_FOLLOWER error. However, if the raft client finishes shutdown before handling the request, the request will not be handled. Shutdown the controller before doing KafkaFuture#get for the client request, so we can make sure the request is handled by another controller eventually.

Signed-off-by: PoAn Yang <payang@apache.org>

Reviewers: Luke Chen <showuon@gmail.com>

lucasbru committed:
KAFKA-18323: Add StreamsGroup class (#18729)

Implements a memory model for representing streams groups in the group coordinator, as well as group count and rebalance metrics.

Reviewers: Bill Bejeck <bill@confluent.io>, Bruno Cadonna <bruno@confluent.io>

Yunyung committed:
KAFKA-18777 add `PartitionsWithLateTransactionsCount` to BrokerMetricNamesTest (#18869)

Rewrite BrokerMetricNamesTest using ReplicaManager.MetricNames, ensuring that all metrics are always included. This helps prevent issues like PartitionsWithLateTransactionsCount not being correctly included in the test before.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

ShivsundarR committed:
KAFKA-18769: Improve leadership changes handling in ShareConsumeRequestManager. (#18851)

Reviewers: Andrew Schofield <aschofield@confluent.io>

pdruley committed:
Merge branch 'apache:trunk' into pdruley/minisrDocChange

Commit:
rewrite of minisr

pdruley commented:
@divijvaidya I somehow managed to add a bunch of commits to this branch that are ahead of trunk. I am not an everyday git user so it's definitely something I did wrong.

I am just going to close this PR and start a new branch with the changes on new config file.
